### PR TITLE
fix race reported by rv-predict

### PR DESCRIPTION
**Pull request by dwightguth**
@yilongli please review. Also, if you can, let me know if you think this is the correct solution to the race. I think it should be fine, because it's a lazy computation and therefore all we really need is to make the put be well ordered and atomic, but I want to get your input.

### CODE REVIEW

File: java-backend/src/main/java/org/kframework/backend/java/kil/KItem.java [hunk 1]
```diff
@@ -126,7 +125,9 @@ private KItem(Term kLabel, Term kList, TermContext termContext, Tool tool, Sourc
                     && definition.sortPredicateRulesOn(kLabelConstant).isEmpty();
             if (enableCache) {
                 cacheTabColKey = new CacheTableColKey(kLabelConstant, (KList) kList);
-                cacheTabVal = definition.getSortCacheTable().get(cacheTabColKey);
+                synchronized(definition.getSortCacheTable()) {
```
Comment (yilongli) [inline java-backend/src/main/java/org/kframework/backend/java/kil/KItem.java]: Is there any reason that we need to expose the `sortCacheTable` in `Definition` with a getter? If not, what about only providing methods to put/get key/value of the cache table and do synchronization inside those methods?

**Comment by yilongli** [comment]
This is the related race report given by rv-predict:

```
Possible data race on field java.util.HashMap.$state: {{{
    Concurrent read in thread T65 (locks held: {})
        at org.kframework.backend.java.kil.KItem.<init>(KItem.java:135)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:172)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:254)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:168)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:254)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:168)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:254)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:168)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:267)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:267)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.transformAndEval(KILtoBackendJavaKILTransformer.java:128)
        at org.kframework.backend.java.symbolic.JavaSymbolicExecutor.search(JavaSymbolicExecutor.java:120)
        at org.kframework.krun.tools.Executor$Tool.search(Executor.java:153)
        at org.kframework.krun.tools.Executor$Tool.run(Executor.java:124)
        at org.kframework.transformation.TransformationCompositionProvider$2.run(TransformationCompositionProvider.java:120)
        at org.kframework.krun.KRunFrontEnd.run(KRunFrontEnd.java:83)
        at org.kframework.main.FrontEnd.main(FrontEnd.java:51)
        at org.kframework.main.Main.runApplication(Main.java:110)
        at org.kframework.kserver.KServerFrontEnd.run(KServerFrontEnd.java:119)
        at org.kframework.main.Main.nailMain(Main.java:71)
    T65 is created by T12
        at com.martiansoftware.nailgun.NGSessionPool.take(NGSessionPool.java:85)

    Concurrent write in thread T53 (locks held: {})
        at org.kframework.backend.java.kil.KItem.computeSort(KItem.java:259)
        at org.kframework.backend.java.kil.KItem.sort(KItem.java:625)
        at org.kframework.backend.java.kil.KItem$CacheTableColKey.<init>(KItem.java:724)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:172)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:254)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:168)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:267)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:267)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.transformAndEval(KILtoBackendJavaKILTransformer.java:128)
        at org.kframework.backend.java.symbolic.JavaSymbolicExecutor.search(JavaSymbolicExecutor.java:120)
        at org.kframework.krun.tools.Executor$Tool.search(Executor.java:153)
        at org.kframework.krun.tools.Executor$Tool.run(Executor.java:124)
        at org.kframework.transformation.TransformationCompositionProvider$2.run(TransformationCompositionProvider.java:120)
        at org.kframework.krun.KRunFrontEnd.run(KRunFrontEnd.java:83)
        at org.kframework.main.FrontEnd.main(FrontEnd.java:51)
        at org.kframework.main.Main.runApplication(Main.java:110)
        at org.kframework.kserver.KServerFrontEnd.run(KServerFrontEnd.java:119)
        at org.kframework.main.Main.nailMain(Main.java:71)
    T53 is created by T12
        at com.martiansoftware.nailgun.NGSessionPool.take(NGSessionPool.java:85)
}}}
```

As pointed out by Dwight, there is a few stack frames missing, e.g., `KItem.of`, in the report. I am looking into that.

**Comment by yilongli** [comment]
Race report after fixing the missing stackframes:

```
Possible data race on field java.util.HashMap.$state: {{{
    Concurrent read in thread T35 (locks held: {})
        at org.kframework.backend.java.kil.KItem.<init>(KItem.java:135)
        at org.kframework.backend.java.kil.KItem.of(KItem.java:88)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:172)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:254)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:168)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:267)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:267)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.transformAndEval(KILtoBackendJavaKILTransformer.java:128)
        at org.kframework.backend.java.symbolic.JavaSymbolicExecutor.search(JavaSymbolicExecutor.java:120)
        at org.kframework.krun.tools.Executor$Tool.search(Executor.java:153)
        at org.kframework.krun.tools.Executor$Tool.run(Executor.java:124)
        at org.kframework.transformation.TransformationCompositionProvider$2.run(TransformationCompositionProvider.java:120)
        at org.kframework.krun.KRunFrontEnd.run(KRunFrontEnd.java:83)
        at org.kframework.main.FrontEnd.main(FrontEnd.java:51)
        at org.kframework.main.Main.runApplication(Main.java:110)
        at org.kframework.kserver.KServerFrontEnd.run(KServerFrontEnd.java:119)
        at org.kframework.main.Main.nailMain(Main.java:71)
    T35 is created by T12
        at com.martiansoftware.nailgun.NGSessionPool.take(NGSessionPool.java:85)

    Concurrent write in thread T16 (locks held: {})
        at org.kframework.backend.java.kil.KItem.computeSort(KItem.java:259)
        at org.kframework.backend.java.kil.KItem.sort(KItem.java:625)
        at org.kframework.backend.java.kil.KItem$CacheTableColKey.<init>(KItem.java:724)
        at org.kframework.backend.java.kil.KItem.<init>(KItem.java:134)
        at org.kframework.backend.java.kil.KItem.of(KItem.java:88)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:172)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:254)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:168)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:267)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.visit(KILtoBackendJavaKILTransformer.java:267)
        at org.kframework.backend.java.symbolic.KILtoBackendJavaKILTransformer.transformAndEval(KILtoBackendJavaKILTransformer.java:128)
        at org.kframework.backend.java.symbolic.JavaSymbolicExecutor.search(JavaSymbolicExecutor.java:120)
        at org.kframework.krun.tools.Executor$Tool.search(Executor.java:153)
        at org.kframework.krun.tools.Executor$Tool.run(Executor.java:124)
        at org.kframework.transformation.TransformationCompositionProvider$2.run(TransformationCompositionProvider.java:120)
        at org.kframework.krun.KRunFrontEnd.run(KRunFrontEnd.java:83)
        at org.kframework.main.FrontEnd.main(FrontEnd.java:51)
        at org.kframework.main.Main.runApplication(Main.java:110)
        at org.kframework.kserver.KServerFrontEnd.run(KServerFrontEnd.java:119)
        at org.kframework.main.Main.nailMain(Main.java:71)
    T16 is created by T12
        at com.martiansoftware.nailgun.NGSessionPool.take(NGSessionPool.java:85)
}}}
```

**Comment by dwightguth** [comment]
@yilongli please re-review

**Comment by dwightguth** [comment]
This is the second time I've seen a transient failure in kool typed static today. @yilongli , any chance you could run rv predict on the test suite for kool and see if you catch any race conditions that look plausible?

**Comment by dwightguth** [comment]
Jenkins: test this please

**Comment by yilongli** [comment]
Yeah, I will put it on top my TODO list.

**Comment by yilongli** [comment]
lgtm

**Pull request merged**